### DISCUSSION TITLE
typescript prints errors if my model has reference and I want to pass primary key

### DISCUSSION THREAD

**Discussion by NaumovEvgeniy** [prisma/prisma#18810] [Q&A] [answered]
### Bug description

I have two models 
```prisma
model User {
  id    Int    @id @default(autoincrement())
  posts Post[]
}

model Post {
  id       Int  @id @default(autoincrement())
  author   User @relation(fields: [authorId], references: [id])
  authorId Int // relation scalar field  (used in the `@relation` attribute above)
}
```

I want to create a post with user. 
```typescript
    this.prisma.post.create({
      data: {
        id: 12,
        author: {
          create: {id: 12}
        }
      }
    })
```

And I've got an error
```
TS2322: Type '{ id: number; author: { create: { id: number; }; }; }' is not assignable to type '(Without<PostCreateInput, PostUncheckedCreateInput> & PostUncheckedCreateInput) | (Without<...> & PostCreateInput)'.
  Type '{ id: number; author: { create: { id: number; }; }; }' is not assignable to type 'Without<PostUncheckedCreateInput, PostCreateInput> & PostCreateInput'.
    Type '{ id: number; author: { create: { id: number; }; }; }' is not assignable to type 'Without<PostUncheckedCreateInput, PostCreateInput>'.
      Types of property 'id' are incompatible.
        Type 'number' is not assignable to type 'never'.

```

it works correctly without passing `id`

### How to reproduce

1. Create `schema.prisma`
2. Write models

```prisma
model User {
  id    Int    @id @default(autoincrement())
  posts Post[]
}

model Post {
  id       Int  @id @default(autoincrement())
  author   User @relation(fields: [authorId], references: [id])
  authorId Int // relation scalar field  (used in the `@relation` attribute above)
}
```

3. generate schema `npx prisma generate --schema ./path/to/your/schema`
4. try to create a post
```typescript
this.prisma.post.create({
      data: {
        id: 12,
        author: {
          create: {id: 12}
        }
      }
    })
```

### Expected behavior

There is no error of typescript

### Prisma information

Schema and client code has written above

### Environment & setup

- OS: Linux fedora 6.0.8-100.fc35.x86_64 
- Database: PostgreSQL
- Node.js version: v18.15.0

### Prisma Version

```
4.13.0
```

**Comment by nurul3101** [comment, marked as answer]
Hello @NaumovEvgeniy 👋 

Based on your schema file and the query you are invoking, it is mandatory to pass an authorId when creating a Post, but in the query
```ts
this.prisma.post.create({
      data: {
        id: 12,
        author: {
          create: {id: 12}
        }
      }
    })
```
You are trying to create a post and an author, but the author is not connected to the post.
Doing something like this would work
```ts
  await prisma.user.create({
    data: {
      id: 1,
    },
  });
  
  await prisma.post.create({
    data: {
      id: 12,
      authorId: 1,
    },
  });
```

**Reply by nurul3101** [reply]
Let us know in case you are still facing this error.

**Reply by nurul3101** [reply]
@NaumovEvgeniy Do let us know in case you need any assistance.

**Reply by NaumovEvgeniy** [reply]
@nurul3101 

Ye, I see your point.

But how can you explain error if I provide an `undefined` or `null` value in the `id` field?

```typescript
    this.prisma.post.create({
      data: {
        id: undefined,
        author: {
          create: {id: 12}
        }
      }
    })

```

Error
```
TS2322: Type '{ id: undefined; author: { create: { id: number; }; }; }' is not assignable to type '(Without<PostCreateInput, PostUncheckedCreateInput> & PostUncheckedCreateInput) | (Without<...> & PostCreateInput)'.
  Type '{ id: undefined; author: { create: { id: number; }; }; }' is not assignable to type 'Without<PostUncheckedCreateInput, PostCreateInput> & PostCreateInput'.
    Type '{ id: undefined; author: { create: { id: number; }; }; }' is not assignable to type 'Without<PostUncheckedCreateInput, PostCreateInput>'.
      Types of property 'id' are incompatible.
        Type 'undefined' is not assignable to type 'never'.
    47 |
    48 |     this.prisma.post.create({
  > 49 |       data: {
       |       ^^^^
    50 |         id: undefined,
    51 |         author: {
    52 |           create: {id: 12}

```

**Reply by nurul3101** [reply]
If you pass the `undefined` value to `id` field, it would be equivalent of passing no `id` at all. This [guide](https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined) explains how null and undefined are treated in prisma queries.

**Reply by nurul3101** [reply]
Let us know in case you have any queries.

**Reply by nurul3101** [reply]
If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by NaumovEvgeniy** [reply]
> If you pass the undefined value to id field, it would be equivalent of passing no id at all.

If this code 
```typescript
    this.prisma.post.create({
      data: {
        author: {
          create: {id: 12}
        }
      }
    })
```


equivalent to

```typescript
    this.prisma.post.create({
      data: {
        id: undefined,
        author: {
          create: {id: 12}
        }
      }
    })
```

why typescript rises an error if I pass an `undefined`  value to the `id` field?

**Reply by nurul3101** [reply]
TypeScript is raising an error because you are explicitly setting the `id` field of a Post to undefined. The id field in the Post model is of type `Int` with the `@default(autoincrement())` directive. This means Prisma will automatically generate an integer value for the id field when a new post is created.

When you explicitly set the id field to undefined, TypeScript is warning you that this is not a valid value according to the schema, as the id field should be of type Int. Prisma will ignore the undefined value and still generate a unique id when creating the Post, but it's not the expected behaviour from a TypeScript perspective.

**Reply by nurul3101** [reply]
Let us know in case you have any further queries.
